README.md
# NVIDIA Power Manager

Indicator applet for laptop users with NVIDIA/Intel hybrid GPUs, allowing one
to manually turn the discrete GPU on or off when needed.

It should work on any Linux distribution that includes the minimum
dependencies listed in the Prerequisites section.

This version uses `bbswitch` to power the NVIDIA GPU off. It doesn't use 
`prime-select` or any other utilities so the primary idea of this applet is
to make it possible to manually turn discrete GPU ON or OFF. It assumes
specific setup when discrete GPU is only used for things like GPGPU and X
server runs on integrated GPU.

This implementation is based on the [NVIDIA Power Indicator](
https://github.com/andrebrait/nvidia-power-indicator) applet by
André Brait Carneiro Fabotti with the following changes:

* Organized as a `python3` package and can be installed using `pip`.
* Indicator icon displays only nVidia icon either color or symbolic -
  corresponds to enabled / disabled GPU.
* Indicator menu simplified and doesn't have option to turn management off.
* Using `py3nvml` module to query list of processes that use GPU (instead of
  invoking `pgrep`)
* Message box displaying information about GPU in use displays list of 
  processes that use GPU.

## Prerequisites

Make sure you have installed and enabled:

* NVIDIA driver, version 331.20 or higher
* `mesa-utils` package
* `python3` package
* `gir1.2-appindicator3` package
* `bbswitch-dkms` or equivalent package

Or simply run the following, which will install all dependencies and the latest NVIDIA driver for your GPU (if it's supported by NVIDIA's latest drivers).

```bash
sudo apt-get install python3 mesa-utils bbswitch-dkms gir1.2-appindicator3
sudo apt-get install $(sudo ubuntu-drivers devices | grep -o nvidia-[[:digit:]]*)
```

## Installation

There are two main ways to install this package: using pip or from Ubuntu PPA.

The first way is more pythonic but may fail to setup package properly. The package needs to install a file in `sudoers.d` and also XDG autostart file. Both are inside of `/etc` directory.  
Depending on the environment and pip version installation to `/etc` may silently fail in install those files to for example `/usr/local/lib/etc`.
This will result in application not starting with graphical session and also failing to switch GPU state if started manually.

The second way installs application as a `.deb` package. The downside is that application depends on `py3nvml` python package and that package is not available in Ubuntu repositories. So the PPA archive includes `python3-py3nvml`
package made from the published PyPI version.

### Install from PPA

Add PPA and update list of available packages:

```bash
sudo add-apt-repository ppa:dm0/nvidia-power-manager
sudo apt-get update
```

Install 

```bash
sudo apt-get install nvidia-power-manager
```

### Install using pip

Use `pip` to install package. This package installs XDG autostart file and 
`sudoers.d` file so `pip install` must be invoked with `sudo`.

```bash
sudo pip3 install git+https://github.com/dm0/nvidia-power-manager#egg=nvidia-power-manager
```

Applet will start on next reboot. Use the following command to start in current session:

```bash
nvidia-power-manager & disown
```

## Troubleshooting

### `appindicator` module missing
Install the `gir1.2-appindicator3` package.

### Couldn't find RGB GLX visual or fbconfig
Install the `mesa-utils` package.

### Installing from root session
Installing from root session in Ubuntu doesn't work. Use `sudo` to install package. 

### Application doesn't auto start when installed with pip
Use `pip3 show -f nvidia-power-manager` to view installed files.  
There should be files installed to `/etc`.

For example (not that path are shown relative to "Location"):
```
$ pip3 show -f nvidia-power-manager
Name: nvidia-power-manager
...
Location: /usr/local/lib/python3.6/dist-packages
Requires: py3nvml
Required-by: 
Files:
  ../../../../../etc/sudoers.d/99-nvidia-power-manager-sudoers
  ../../../../../etc/xdg/autostart/nvidia-power-manager.desktop
  ../../../../lib/nvidia-power-manager/gpuswitcher
  ../../../bin/nvidia-power-manager
  ../../../share/icons/nvidia-power-manager-active.svg
...
```

If files are installed to a wrong location you man fix it manually or install package from PPA.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
Unreleased changes will be added to this section first.


## [1.1.3] - 2019-09-29
Note: version 1.1.2 skipped due to a wrong build was submitted with this number.
### Added
- Fixed error messages in syslog caused by a wrong modprobe / bbswitch order.

[1.1.3]: https://github.com/dm0/nvidia-power-manager/compare/v1.1.1...v1.1.3


## [1.1.1] - 2019-06-15
### Added
- Added support for building debian package.

[1.1.1]: https://github.com/dm0/nvidia-power-manager/compare/v1.1.0...v1.1.1


## [1.1.0] - 2018-08-11
### Added
- Added support for logging. Logs to syslog by default. Can log to `stderr` if started with `--debug` option.

### Changed
- Improved error reporting.

### Fixed
- Fixed compatibility with Ubuntu 18.04. 

[1.1.0]: https://github.com/dm0/nvidia-power-manager/compare/v1.0.2...v1.1.0


## [1.0.2] - 2018-05-21
### Changed
- Readme improvements: added information about starting application manually.

[1.0.2]: https://github.com/dm0/nvidia-power-manager/compare/v1.0.1...v1.0.2


## [1.0.1] - 2018-05-21
### Changed
- Added python package meta-data: project URL and license.

### Removed
- `.idea/` files that were in the forked repository.

[1.0.1]: https://github.com/dm0/nvidia-power-manager/compare/v1.0.0...v1.0.1


## [1.0.0] - 2018-05-21
Initial release.

[1.0.0]: https://github.com/dm0/nvidia-power-manager/releases/tag/v1.0.0


nvidia_power_indicator/nvidia_power_indicator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# This file is part of NVIDIA Power Manager - indicator applet for NVIDIA
# laptops.
# Copyright (C) 2018 Dmitri Lapin
#
# This work is based on the NVIDIA Power Indicator applet by
# André Brait Carneiro Fabotti
#
# NVIDIA Power Manager is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# NVIDIA Power Manager is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NVIDIA Power Manager. If not, see <http://www.gnu.org/licenses/>.

import sys
import os
import signal
import subprocess

from py3nvml import py3nvml

import gi.repository

gi.require_version('Gtk', '3.0')
gi.require_version('AppIndicator3', '0.1')
from gi.repository import Gtk
from gi.repository import AppIndicator3

from threading import Timer

from typing import Optional, List
from collections import namedtuple

ProcInfo = namedtuple('ProcInfo', ['name', 'pid', 'mem'])


class Indicator:
    APP_NAME = 'NVIDIA Power Manager'
    LIB_PATH = '/usr/lib/nvidia-power-manager/'
    SCRIPT_CMD = 'sudo ' + LIB_PATH + 'gpuswitcher'
    BASEDIR = os.path.dirname(os.path.realpath(__file__))
    ACTIVE_ICON = os.path.join(BASEDIR, 'icons', 'active.svg')
    INACTIVE_ICON = os.path.join(BASEDIR, 'icons', 'inactive.svg')

    def __init__(self):
        self.menu = Gtk.Menu()

        self.switch_power_management = Gtk.MenuItem()
        self.switch_power_management.connect('activate', self.switch_nv_power)
        self.menu.append(self.switch_power_management)

        self.set_nv_pm_labels()

        item = Gtk.SeparatorMenuItem()
        item.show()
        self.menu.append(item)

        item = Gtk.MenuItem('Quit')
        item.connect('activate', self.terminate)
        item.show()
        self.menu.append(item)

        self.icon = AppIndicator3.Indicator.new(
            Indicator.APP_NAME,
            '',
            AppIndicator3.IndicatorCategory.APPLICATION_STATUS
        )
        self.icon.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        self.icon.set_menu(self.menu)

        self.t = None
        self.refresh()

    def terminate(self, window=None, data=None) -> None:
        self.t.cancel()
        Gtk.main_quit()

    def execute(self):
        Gtk.main()

    def nv_power_switch_string(self, nvidia_on) -> str:
        if nvidia_on is None:
            nvidia_on = self.is_nvidia_on()
        return 'Power {state} GPU'.format(
            state=('Off' if nvidia_on else 'On'))

    def is_nvidia_on(self) -> bool:
        with open('/proc/acpi/bbswitch', 'r') as fh:
            out = fh.readline()
        return out.strip().lower().endswith('on')

    def switch_nv_power(self, dude) -> None:
        if self.is_nvidia_on():
            self.turn_nv_off()
        else:
            self.turn_nv_on()

    def turn_nv_on(self) -> None:
        os.system(Indicator.SCRIPT_CMD + ' on')
        self.set_nv_pm_labels()

    def get_device_procs(self, device_id: int) -> Optional[List[ProcInfo]]:
        py3nvml.nvmlInit()
        dev_count = py3nvml.nvmlDeviceGetCount()  # type: int
        if not (0 <= device_id < dev_count):
            raise RuntimeError('Failed to query GPU with nvml')
        handle = py3nvml.nvmlDeviceGetHandleByIndex(device_id)
        result = []
        try:
            for proc in py3nvml.nvmlDeviceGetComputeRunningProcesses(handle):
                try:
                    name = str(py3nvml.nvmlSystemGetProcessName(proc.pid))
                except py3nvml.NVMLError as err:
                    if (err.value == py3nvml.NVML_ERROR_NOT_FOUND):
                        # exited?
                        continue
                    raise
                mem = proc.usedGpuMemory / 1024 / 1024
                result.append(ProcInfo(name, proc.pid, mem))
        finally:
            py3nvml.nvmlShutdown()

        return result

    def turn_nv_off(self) -> None:
        message = None
        try:
            procs = self.get_device_procs(0)
        except Exception as e:
            procs = None
            message = 'Failed to query device: {error}'.format(error=e)
        else:
            if len(procs) > 0:
                proc_info = [
                    ('{proc.name} (pid: {proc.pid}, mem: '
                     '{proc.mem:0.1}MB').format(proc=p) for p in procs
                ]
                message = (
                    'It seems there are programs using the NVIDIA GPU. '
                    'They need to be stopped before turning the GPU off:\n'
                ) + proc_info.join('\n')
        if message is not None:
            dialog = Gtk.MessageDialog(None,
                                       Gtk.DialogFlags.MODAL,
                                       Gtk.MessageType.ERROR,
                                       Gtk.ButtonsType.OK,
                                       message)
            dialog.set_deletable(False)
            dialog.run()
        else:
            os.system(Indicator.SCRIPT_CMD + ' off')
            self.set_nv_pm_labels()

    def set_nv_pm_labels(self, nvidia_on=None) -> None:
        self.switch_power_management.set_label(
            self.nv_power_switch_string(nvidia_on))
        self.switch_power_management.show()

    def ignore(self, *args):
        return Gtk.ResponseType.CANCEL

    def refresh(self) -> None:
        self.t = Timer(2, self.refresh)
        self.t.start()

        nvidia_on = self.is_nvidia_on()
        self.icon.set_icon(
            Indicator.ACTIVE_ICON if nvidia_on else Indicator.INACTIVE_ICON)
        self.icon.set_title(
            'GPU powered ON' if nvidia_on else 'GPU powered OFF')
        self.set_nv_pm_labels(nvidia_on)


def kill_other_instances() -> None:
    otherpid = subprocess.run(['pgrep', '-f', 'nvidia-power-manager'])
    if otherpid.returncode == 0:
        otherpid = str(otherpid.stdout)  # type:str
        pidlist = otherpid.splitlines()  # type:List[str]
        for pid in pidlist:
            if pid and pid.isnumeric():
                pid = int(pid)  # type:int
                if pid != os.getpid():
                    try:
                        os.kill(pid, signal.SIGTERM)
                        os.kill(pid, signal.SIGKILL)
                    except ProcessLookupError:
                        pass


if __name__ == '__main__':

    # If bbswitch isn't installed or isn't supported, exit cleanly
    if not os.path.isfile('/proc/acpi/bbswitch'):
        sys.exit(0)

    kill_other_instances()
    Indicator().execute()


nvidia_power_manager/app.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# This file is part of NVIDIA Power Manager - indicator applet for NVIDIA
# laptops.
# Copyright (C) 2018 Dmitri Lapin
#
# This work is based on the NVIDIA Power Indicator applet by
# André Brait Carneiro Fabotti
#
# NVIDIA Power Manager is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# NVIDIA Power Manager is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NVIDIA Power Manager. If not, see <http://www.gnu.org/licenses/>.

"""
NVidia Power Manager indicator applet.
"""

import sys
import os
import signal
import subprocess
import argparse
import logging
import logging.handlers

from py3nvml import py3nvml

import gi.repository

gi.require_version('Gtk', '3.0')
gi.require_version('AppIndicator3', '0.1')
from gi.repository import Gtk
from gi.repository import AppIndicator3

from threading import Timer

from typing import Optional, List
from collections import namedtuple

ProcInfo = namedtuple('ProcInfo', ['name', 'pid', 'mem'])

logger = logging.getLogger('nvidia-power-manager')


class Indicator:
    """
    Indicator application.

    Attributes
    ----------
    ACTIVE_ICON : str
        Path to "active" state icon (GPU switched on)
    APP_NAME : str
        Application name
    BASEDIR : str
        Base directory for relative paths
    icon : AppIndicator3.Indicator
        Indicator icon instance
    INACTIVE_ICON : str
        Path to "inactive" state icon (GPU switched off)
    LIB_PATH : str
        Path to installed scripts directory
    menu : Gtk.Menu
        GTK Menu instance (popup menu)
    SCRIPT_CMD : str
        gpuswitcher script command line
    switch_power_management : Gtk.MenuItem
        Menu item to toggle GPU state
    t : threading.Timer
        Timer used to refresh device state
    """

    APP_NAME = 'NVIDIA Power Manager'
    LIB_PATH = '/usr/lib/nvidia-power-manager/'
    SCRIPT_CMD = 'sudo ' + LIB_PATH + 'gpuswitcher'
    BASEDIR = os.path.dirname(os.path.realpath(__file__))
    ACTIVE_ICON = os.path.join(
        BASEDIR, 'icons', 'nvidia-power-manager-active.svg')
    INACTIVE_ICON = os.path.join(
        BASEDIR, 'icons', 'nvidia-power-manager-inactive.svg')

    def __init__(self):
        """
        Initialize application.
        """
        self.menu = Gtk.Menu()

        self.switch_power_management = Gtk.MenuItem()
        self.switch_power_management.connect('activate', self.switch_nv_power)
        self.menu.append(self.switch_power_management)

        self.set_nv_pm_labels()

        item = Gtk.SeparatorMenuItem()
        item.show()
        self.menu.append(item)

        item = Gtk.MenuItem('Quit')
        item.connect('activate', self.terminate)
        item.show()
        self.menu.append(item)

        self.icon = AppIndicator3.Indicator.new(
            Indicator.APP_NAME,
            '',
            AppIndicator3.IndicatorCategory.APPLICATION_STATUS
        )
        self.icon.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        self.icon.set_menu(self.menu)

        self.t = None
        self.refresh()

    def terminate(self, window=None, data=None) -> None:
        """
        Shutdown application.

        Parameters
        ----------
        window : unused
            Placeholder for the actual parameter passed by GTK
        data : unused
            Placeholder for the actual parameter passed by GTK
        """
        self.t.cancel()
        Gtk.main_quit()

    def execute(self):
        """
        Enter application message loop.
        """
        Gtk.main()

    def nv_power_switch_string(self, nvidia_on) -> str:
        """
        Return string describing GPU toggle action.

        Parameters
        ----------
        nvidia_on : bool
            Current GPU state.

        Returns
        -------
        str
            String describing current GPU toggle action.
        """
        if nvidia_on is None:
            nvidia_on = self.is_nvidia_on()
        return 'Power {state} GPU'.format(
            state=('Off' if nvidia_on else 'On'))

    def is_nvidia_on(self) -> bool:
        """
        Check current GPU state.

        Returns
        -------
        bool
            Returns True if GPU is enabled and False otherwise.
        """
        with open('/proc/acpi/bbswitch', 'r') as fh:
            out = fh.readline()
        logger.debug('is_nvidia_on: bbswitch state: %s', out)
        return out.strip().lower().endswith('on')

    def switch_nv_power(self, dude) -> None:
        """
        Toggle GPU state.

        Parameters
        ----------
        dude : unused
            Placeholder for the actual parameter passed by GTK.
        """
        if self.is_nvidia_on():
            self.turn_nv_off()
        else:
            self.turn_nv_on()

    def turn_nv_on(self) -> None:
        """
        Switch on GPU.
        """
        logger.info('Swithing GPU on...')
        logger.debug('turn_nv_on: executing "%s on"', Indicator.SCRIPT_CMD)
        os.system(Indicator.SCRIPT_CMD + ' on')
        self.set_nv_pm_labels()

    def get_device_procs(self, device_id: int) -> Optional[List[ProcInfo]]:
        """
        List processes running on the GPU.

        Parameters
        ----------
        device_id : int
            Device identifier

        Returns
        -------
        Optional[List[ProcInfo]]
            List of ProcInfo named tuples (name, pid, mem fields)

        Raises
        ------
        RuntimeError
            In case of py3nvml failure.
        """
        py3nvml.nvmlInit()
        dev_count = py3nvml.nvmlDeviceGetCount()  # type: int
        if not (0 <= device_id < dev_count):
            raise RuntimeError('Failed to query GPU with nvml')
        handle = py3nvml.nvmlDeviceGetHandleByIndex(device_id)
        result = []
        try:
            for proc in py3nvml.nvmlDeviceGetComputeRunningProcesses(handle):
                try:
                    name = str(py3nvml.nvmlSystemGetProcessName(proc.pid))
                except py3nvml.NVMLError as err:
                    if (err.value == py3nvml.NVML_ERROR_NOT_FOUND):
                        # exited?
                        continue
                    raise
                mem = proc.usedGpuMemory / 1024 / 1024
                result.append(ProcInfo(name, proc.pid, mem))
        finally:
            py3nvml.nvmlShutdown()

        return result

    def turn_nv_off(self) -> None:
        """
        Switch off GPU device.
        """
        logger.info('Swithing GPU off...')
        message = None
        message_details = None
        try:
            procs = self.get_device_procs(0)
        except Exception as e:
            procs = None
            message = 'Failed to query device: {error}'.format(error=e)
            logger.exception('Failed to query device')
        else:
            if len(procs) > 0:
                proc_info = [
                    ('{proc.name} (pid: {proc.pid}, mem: '
                     '{proc.mem:0.1f} MB)').format(proc=p) for p in procs
                ]
                message = (
                    'It seems there are programs using the NVIDIA GPU. '
                    'They need to be stopped before turning the GPU off\n'
                )
                message_details = '\n'.join(proc_info)
                logger.warning('GPU device is in use. Can\'t stop')
        if message is not None:
            dialog = Gtk.MessageDialog(Gtk.Window(),
                                       Gtk.DialogFlags.MODAL,
                                       Gtk.MessageType.ERROR,
                                       Gtk.ButtonsType.OK,
                                       message)
            if message_details is not None:
                dialog.format_secondary_text(message_details)
            dialog.run()
            dialog.destroy()
        else:
            os.system(Indicator.SCRIPT_CMD + ' off')
            self.set_nv_pm_labels()

    def set_nv_pm_labels(self, nvidia_on=None) -> None:
        """
        Update menu item label.

        Parameters
        ----------
        nvidia_on : bool, optional
            Current GPU state.
        """
        self.switch_power_management.set_label(
            self.nv_power_switch_string(nvidia_on))
        self.switch_power_management.show()

    def ignore(self, *args):
        """
        Handle ignore.

        Parameters
        ----------
        *args
            Unused.

        Returns
        -------
        Gtk.ResponseType
            Gtk.ResponseType.CANCEL
        """
        return Gtk.ResponseType.CANCEL

    def refresh(self) -> None:
        """
        Update device status.
        """
        logger.debug('refresh: updating device state')
        self.t = Timer(2, self.refresh)
        self.t.start()

        nvidia_on = self.is_nvidia_on()
        self.icon.set_icon(
            Indicator.ACTIVE_ICON if nvidia_on else Indicator.INACTIVE_ICON)
        self.icon.set_title(
            'GPU powered ON' if nvidia_on else 'GPU powered OFF')
        self.set_nv_pm_labels(nvidia_on)


def kill_other_instances() -> None:
    """
    Kill other instances of the application.
    """
    logger.info('Killing other instances')
    otherpid = subprocess.run(['pgrep', '-f', 'nvidia-power-manager'],
                              stdout=subprocess.PIPE)
    if otherpid.returncode == 0:
        otherpid = str(otherpid.stdout)  # type:str
        pidlist = otherpid.splitlines()  # type:List[str]
        for pid in pidlist:
            if pid and pid.isnumeric():
                pid = int(pid)  # type:int
                if pid != os.getpid():
                    try:
                        logger.info('Killing PID %d', pid)
                        os.kill(pid, signal.SIGTERM)
                        os.kill(pid, signal.SIGKILL)
                    except ProcessLookupError:
                        pass


def main():
    """
    Application entry point.
    """
    parser = argparse.ArgumentParser(
        description='NVidia Power Manager applet',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    parser.add_argument('--log-level', '-l', default='WARN',
                        choices=['DEBUG', 'WARN', 'ERROR'],
                        help='Log level')
    parser.add_argument('--syslog-socket', default='/dev/log',
                        help='Syslog unix socket path')
    parser.add_argument('--debug', action='store_true',
                        help='Enable debug mode: set log level to DEBUG and '
                        'log to stderr')

    args = parser.parse_args()

    if args.debug:
        args.log_level = 'DEBUG'
        log_handler = logging.StreamHandler()
        log_formatter = logging.Formatter(
            '%(asctime)s: %(levelname)s: %(message)s')
    else:
        log_handler = logging.handlers.SysLogHandler(args.syslog_socket)
        log_formatter = logging.Formatter(
            '%(name)s: %(levelname)s: %(message)s')

    logger.setLevel(getattr(logging, args.log_level.upper()))
    log_handler.setFormatter(log_formatter)
    logger.addHandler(log_handler)
    logging.captureWarnings(True)

    # If bbswitch isn't installed or isn't supported, exit cleanly
    if not os.path.isfile('/proc/acpi/bbswitch'):
        logger.critical('Check bbswitch installation (/proc/acpi/bbswitch '
                        'doesn\'t exist on inaccessible)')
        sys.exit(0)

    kill_other_instances()
    Indicator().execute()


if __name__ == '__main__':
    main()
